# **Understanding Dataset**

In [2]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

In [3]:
loyalty_df = pd.read_csv("Customer Loyalty History.csv")
flight_df = pd.read_csv("Customer Flight Activity.csv")
calendar_df = pd.read_csv("Calendar.csv")

In [4]:
print("Loyalty Dataset Shape :", loyalty_df.shape)
print("Flight Dataset Shape  :", flight_df.shape)
print("Calendar Dataset Shape:", calendar_df.shape)

Loyalty Dataset Shape : (16737, 16)
Flight Dataset Shape  : (392936, 8)
Calendar Dataset Shape: (2557, 4)


In [5]:
loyalty_df.head()

,Loyalty Number,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month
0,480934,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,NaN,NaN
1,549612,Canada,Alberta,Edmonton,T3G 6Y6,Male,College,NaN,Divorced,Star,3839.61,Standard,2016,3,NaN,NaN
2,429460,Canada,British Columbia,Vancouver,V6E 3D9,Male,College,NaN,Single,Star,3839.75,Standard,2014,7,2018.0,1.0
3,608370,Canada,Ontario,Toronto,P1W 1K4,Male,College,NaN,Single,Star,3839.75,Standard,2013,2,NaN,NaN
4,530508,Canada,Quebec,Hull,J8Y 3Z5,Male,Bachelor,103495.0,Married,Star,3842.79,Standard,2014,10,NaN,NaN


In [6]:
flight_df.head()

,Loyalty Number,Year,Month,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
0,100590,2018,6,12,15276,22914.0,0,0
1,100590,2018,7,12,9168,13752.0,0,0
2,100590,2018,5,4,6504,9756.0,0,0
3,100590,2018,10,0,0,0.0,512,92
4,100590,2018,2,0,0,0.0,0,0


In [7]:
calendar_df.head()

,Date,Start of Year,Start of Quarter,Start of Month
0,2012-01-01,2012-01-01,2012-01-01,2012-01-01
1,2012-01-02,2012-01-01,2012-01-01,2012-01-01
2,2012-01-03,2012-01-01,2012-01-01,2012-01-01
3,2012-01-04,2012-01-01,2012-01-01,2012-01-01
4,2012-01-05,2012-01-01,2012-01-01,2012-01-01


In [8]:
print("LOYALTY DATASET COLUMNS\n")
print(loyalty_df.columns)

print("\n\nFLIGHT DATASET COLUMNS\n")
print(flight_df.columns)

print("\n\nCALENDAR DATASET COLUMNS\n")
print(calendar_df.columns)

LOYALTY DATASET COLUMNS

Index(['Loyalty Number', 'Country', 'Province', 'City', 'Postal Code',
       'Gender', 'Education', 'Salary', 'Marital Status', 'Loyalty Card',
       'CLV', 'Enrollment Type', 'Enrollment Year', 'Enrollment Month',
       'Cancellation Year', 'Cancellation Month'],
      dtype='object')


FLIGHT DATASET COLUMNS

Index(['Loyalty Number', 'Year', 'Month', 'Total Flights', 'Distance',
       'Points Accumulated', 'Points Redeemed', 'Dollar Cost Points Redeemed'],
      dtype='object')


CALENDAR DATASET COLUMNS

Index(['Date', 'Start of Year', 'Start of Quarter', 'Start of Month'], dtype='object')


In [9]:
loyalty_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16737 entries, 0 to 16736
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Loyalty Number      16737 non-null  int64  
 1   Country             16737 non-null  object 
 2   Province            16737 non-null  object 
 3   City                16737 non-null  object 
 4   Postal Code         16737 non-null  object 
 5   Gender              16737 non-null  object 
 6   Education           16737 non-null  object 
 7   Salary              12499 non-null  float64
 8   Marital Status      16737 non-null  object 
 9   Loyalty Card        16737 non-null  object 
 10  CLV                 16737 non-null  float64
 11  Enrollment Type     16737 non-null  object 
 12  Enrollment Year     16737 non-null  int64  
 13  Enrollment Month    16737 non-null  int64  
 14  Cancellation Year   2067 non-null   float64
 15  Cancellation Month  2067 non-null   float64
dtypes: f

In [10]:
flight_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 392936 entries, 0 to 392935
Data columns (total 8 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   Loyalty Number               392936 non-null  int64  
 1   Year                         392936 non-null  int64  
 2   Month                        392936 non-null  int64  
 3   Total Flights                392936 non-null  int64  
 4   Distance                     392936 non-null  int64  
 5   Points Accumulated           392936 non-null  float64
 6   Points Redeemed              392936 non-null  int64  
 7   Dollar Cost Points Redeemed  392936 non-null  int64  
dtypes: float64(1), int64(7)
memory usage: 24.0 MB


In [11]:
loyalty_df.isnull().sum()

Loyalty Number            0
Country                   0
Province                  0
City                      0
Postal Code               0
Gender                    0
Education                 0
Salary                 4238
Marital Status            0
Loyalty Card              0
CLV                       0
Enrollment Type           0
Enrollment Year           0
Enrollment Month          0
Cancellation Year     14670
Cancellation Month    14670
dtype: int64

In [12]:
flight_df.isnull().sum()

Loyalty Number                 0
Year                           0
Month                          0
Total Flights                  0
Distance                       0
Points Accumulated             0
Points Redeemed                0
Dollar Cost Points Redeemed    0
dtype: int64

In [13]:
print("Loyalty duplicates :", loyalty_df.duplicated().sum())
print("Flight duplicates  :", flight_df.duplicated().sum())

Loyalty duplicates : 0


Flight duplicates  : 1922


In [14]:
merged_df = pd.merge(
    flight_df,
    loyalty_df,
    on="Loyalty Number",
    how="left"
)

In [15]:
merged_df.head()

,Loyalty Number,Year,Month,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month
0,100590,2018,6,12,15276,22914.0,0,0,Canada,British Columbia,Vancouver,V6E 3Z3,Female,Bachelor,69841.0,Married,Star,34090.04,2018 Promotion,2018,3,NaN,NaN
1,100590,2018,7,12,9168,13752.0,0,0,Canada,British Columbia,Vancouver,V6E 3Z3,Female,Bachelor,69841.0,Married,Star,34090.04,2018 Promotion,2018,3,NaN,NaN
2,100590,2018,5,4,6504,9756.0,0,0,Canada,British Columbia,Vancouver,V6E 3Z3,Female,Bachelor,69841.0,Married,Star,34090.04,2018 Promotion,2018,3,NaN,NaN
3,100590,2018,10,0,0,0.0,512,92,Canada,British Columbia,Vancouver,V6E 3Z3,Female,Bachelor,69841.0,Married,Star,34090.04,2018 Promotion,2018,3,NaN,NaN
4,100590,2018,2,0,0,0.0,0,0,Canada,British Columbia,Vancouver,V6E 3Z3,Female,Bachelor,69841.0,Married,Star,34090.04,2018 Promotion,2018,3,NaN,NaN


In [16]:
merged_df.groupby("Loyalty Card")[[
    "CLV",
    "Total Flights",
    "Distance",
    "Points Accumulated"
]].mean().sort_values("CLV")

,CLV,Total Flights,Distance,Points Accumulated
Loyalty Card,,,,
Star,6720.538058,1.293715,1930.817111,2012.630477
Nova,8061.795325,1.296792,1947.382081,2036.261847
Aurora,10666.502922,1.294341,1955.074153,2044.264856


In [17]:
merged_df.groupby("Enrollment Type")[[
    "CLV",
    "Total Flights",
    "Points Accumulated"
]].mean()

,CLV,Total Flights,Points Accumulated
Enrollment Type,,,
2018 Promotion,8065.126516,4.180048,9413.351015
Standard,7987.493043,1.213844,1819.694547


In [18]:
merged_df.describe()

,Loyalty Number,Year,Month,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed,Salary,CLV,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month
count,392936.000000,392936.000000,392936.000000,392936.000000,392936.000000,392936.000000,392936.000000,392936.000000,293371.000000,392936.000000,392936.000000,392936.000000,48556.000000,48556.000000
mean,550527.519034,2017.513661,6.513661,1.294888,1941.440201,2027.172345,31.304263,5.635661,79518.020561,7989.614185,2015.163813,6.784728,2016.452920,6.852418
std,258604.580187,0.499814,3.445428,1.962675,3239.975889,3872.139841,126.653775,22.801167,34219.141062,6840.750456,1.947571,3.387459,1.379785,3.424392
min,100018.000000,2017.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,-58486.000000,1898.010000,2012.000000,1.000000,2013.000000,1.000000
25%,327688.000000,2017.000000,4.000000,0.000000,0.000000,0.000000,0.000000,0.000000,59429.000000,3985.320000,2013.000000,4.000000,2016.000000,4.000000
50%,551833.000000,2018.000000,7.000000,0.000000,0.000000,0.000000,0.000000,0.000000,73695.000000,5771.610000,2015.000000,7.000000,2017.000000,7.000000
75%,772194.000000,2018.000000,10.000000,2.000000,3018.000000,3039.000000,0.000000,0.000000,88792.000000,8940.580000,2017.000000,10.000000,2018.000000,10.000000
max,999986.000000,2018.000000,12.000000,28.000000,67284.000000,100926.000000,876.000000,158.000000,407228.000000,83325.380000,2018.000000,12.000000,2018.000000,12.000000


In [19]:
important_cols = [
    "Loyalty Card",
    "Enrollment Type",
    "Gender",
    "Marital Status",
    "Education"
]

for col in important_cols:
    print(f"\n{col}")
    print(merged_df[col].unique())


Loyalty Card
['Star' 'Aurora' 'Nova']

Enrollment Type
['2018 Promotion' 'Standard']

Gender
['Female' 'Male']

Marital Status
['Married' 'Single' 'Divorced']

Education
['Bachelor' 'College' 'Doctor' 'Master' 'High School or Below']


In [20]:
customer_summary = merged_df.groupby("Loyalty Number").agg({
    "Total Flights": "sum",
    "Distance": "sum",
    "Points Accumulated": "sum",
    "Points Redeemed": "sum",
    "Dollar Cost Points Redeemed": "sum",
    "CLV": "first",
    "Loyalty Card": "first",
    "Enrollment Type": "first"
}).reset_index()

In [21]:
customer_summary.head(10)

,Loyalty Number,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed,CLV,Loyalty Card,Enrollment Type
0,100018,46,81190,81190.0,1513,272,7919.20,Aurora,Standard
1,100102,51,68918,68918.0,1195,215,2887.74,Nova,Standard
2,100140,47,72856,72856.0,593,107,2838.07,Nova,Standard
3,100214,22,38236,38236.0,861,155,4170.57,Star,Standard
4,100272,37,54997,54997.0,1007,182,6622.05,Star,Standard
5,100301,41,62849,62849.0,1381,249,48356.96,Nova,Standard
6,100364,33,45955,45955.0,559,101,5143.88,Nova,Standard
7,100380,33,55270,55270.0,1142,206,2465.62,Star,Standard
8,100428,35,58281,58281.0,1658,298,5845.43,Aurora,Standard
9,100504,19,27837,27837.0,0,0,8807.61,Nova,Standard


In [22]:
customer_summary.groupby("Loyalty Card")[[
    "Total Flights",
    "Distance",
    "Points Accumulated",
    "CLV"
]].mean().sort_values("CLV")

,Total Flights,Distance,Points Accumulated,CLV
Loyalty Card,,,,
Star,30.254157,45153.107503,47066.353280,6741.761372
Nova,30.473814,45762.277023,47850.896667,8045.615995
Aurora,30.603675,46226.186060,48335.029892,10672.686325


In [23]:
customer_summary["Redemption Ratio"] = np.where(
    customer_summary["Points Accumulated"] > 0,
    customer_summary["Points Redeemed"] / customer_summary["Points Accumulated"],
    0
)

customer_summary[[
    "Loyalty Card",
    "Redemption Ratio"
]].groupby("Loyalty Card").mean()

,Redemption Ratio
Loyalty Card,
Aurora,0.015736
Nova,0.016453
Star,0.016104


In [24]:
customer_summary.sort_values(
    by="CLV",
    ascending=False
).head(10)

,Loyalty Number,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed,CLV,Loyalty Card,Enrollment Type,Redemption Ratio
10297,652627,0,0,0.0,0,0,83325.38,Star,Standard,0.000000
9611,615459,52,82168,123252.0,631,114,83325.38,Star,2018 Promotion,0.005120
13888,844145,0,0,0.0,0,0,74228.52,Nova,Standard,0.000000
12631,776187,50,61595,61595.0,1529,276,74228.52,Star,Standard,0.024823
9156,592003,38,58764,88146.0,0,0,73225.96,Aurora,2018 Promotion,0.000000
12468,767366,8,15661,15661.0,504,91,73225.96,Aurora,Standard,0.032182
10820,680886,59,86072,86072.0,1645,296,67907.27,Star,Standard,0.019112
13776,838263,0,0,0.0,0,0,67907.27,Aurora,Standard,0.000000
1465,179870,33,48480,48480.0,0,0,66025.75,Aurora,Standard,0.000000
7357,495253,34,49020,49020.0,2611,471,66025.75,Nova,Standard,0.053264


### **Observations**
- Huge Class Imbalance in Cancellation
  - Only 12% customers formally cancelled
  - **Focus on behavioral churn instead of cancellation-only churn**
- Promotion Customers Behave very Differently
  - More avg flights, CLV and points accumulated
  - Promotional users are highly active initially
  - **Analyze if they remain loyal long-term?**
- Strange Salary Values
  - Negative salary is impossible
  - Data anomaly
- Duplicate Flight Rows Exist (1922 duplicates)
  - **Need to decide whether to remove or investigate duplicates**
- **IMPORTANT Insight about CLV**
  - Some users have huge CLV but zero flights/activity
  - Possible Interpretations: previously valuable but inactive, historical high-value customer, silent-churn candidate: dormant high-value cutomers
  - This supports the fact that historical CLV dosen't imply future engagement
- **Exit Behaviour**
  - Some cutomers have 0 recent flights but redeemed points
  - So they used old accumulated points despite not flying recetnly
  - "Let me use remaining points before leaving"

# **Understanding Customer Behaviour**

In [25]:
loyalty_df[loyalty_df["Salary"] < 0]

,Loyalty Number,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month
1082,542976,Canada,Quebec,Montreal,H2Y 4R4,Male,High School or Below,-49830.0,Divorced,Star,24127.50,2018 Promotion,2018,3,NaN,NaN
1894,959977,Canada,British Columbia,Vancouver,V5R 1W3,Female,Bachelor,-12497.0,Married,Aurora,9453.00,2018 Promotion,2018,3,NaN,NaN
2471,232755,Canada,British Columbia,Vancouver,V1E 4R6,Female,Bachelor,-46683.0,Single,Nova,4787.81,2018 Promotion,2018,3,NaN,NaN
3575,525245,Canada,British Columbia,Victoria,V10 6T5,Male,Bachelor,-45962.0,Married,Star,2402.33,2018 Promotion,2018,3,NaN,NaN
3932,603070,Canada,British Columbia,West Vancouver,V6V 8Z3,Female,Bachelor,-19325.0,Single,Star,2893.74,2018 Promotion,2018,3,NaN,NaN
4712,491242,Canada,British Columbia,Dawson Creek,U5I 4F1,Male,Bachelor,-43234.0,Married,Star,7597.91,2018 Promotion,2018,3,NaN,NaN
6560,115505,Canada,Newfoundland,St. John's,A1C 6H9,Male,Bachelor,-10605.0,Married,Nova,5860.17,2018 Promotion,2018,4,NaN,NaN
6570,430398,Canada,Newfoundland,St. John's,A1C 6H9,Male,Bachelor,-17534.0,Married,Nova,49423.80,2018 Promotion,2018,3,NaN,NaN
7373,152016,Canada,Ontario,Toronto,P1J 8T7,Female,Bachelor,-58486.0,Married,Aurora,5067.21,2018 Promotion,2018,2,NaN,NaN
8576,194065,Canada,Ontario,Sudbury,M5V 1G5,Female,Bachelor,-31911.0,Married,Nova,2888.85,2018 Promotion,2018,2,NaN,NaN


In [26]:
missing_salary_pct = (
    loyalty_df["Salary"].isnull().mean()
) * 100

print(f"Missing Salary %: {missing_salary_pct:.2f}%")

missing_salary_count = (
    loyalty_df["Salary"].isnull().sum()
)

print(f"Missing Salary Count: {missing_salary_count}")

Missing Salary %: 25.32%
Missing Salary Count: 4238


In [27]:
loyalty_df["Salary Missing"] = loyalty_df["Salary"].isnull().astype(int)

In [28]:
loyalty_df.loc[
    loyalty_df["Salary"] < 0,
    "Salary"
] = np.nan

In [29]:
loyalty_df["Salary"] = loyalty_df.groupby(
    "Education"
)["Salary"].transform(
    lambda x: x.fillna(x.median())
)

c:\Users\abc\AppData\Local\Programs\Python\Python313\Lib\site-packages\numpy\lib\_nanfunctions_impl.py:1241: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)


Salary missingness was likely behaviorally meaningful rather than random. We preserved this signal using a missing-value indicator and performed education-group-wise imputation followed by global imputation for remaining values to reduce distortion.

In [30]:
missing_salary_count = (
    loyalty_df["Salary"].isnull().sum()
)

print(f"Missing Salary Count: {missing_salary_count}")

Missing Salary Count: 4238


In [31]:
loyalty_df.groupby("Education")["Salary"].apply(
    lambda x: x.isnull().sum()
)

Education
Bachelor                   0
College                 4238
Doctor                     0
High School or Below       0
Master                     0
Name: Salary, dtype: int64

In [32]:
loyalty_df.groupby("Education")["Salary"].median()

Education
Bachelor                 72026.0
College                      NaN
Doctor                  182143.5
High School or Below     61928.0
Master                  105487.0
Name: Salary, dtype: float64

In [33]:
loyalty_df["Salary"] = loyalty_df["Salary"].fillna(
    loyalty_df["Salary"].median()
)

In [34]:
missing_salary_count = (
    loyalty_df["Salary"].isnull().sum()
)

print(f"Missing Salary Count: {missing_salary_count}")

Missing Salary Count: 0


In [35]:
duplicate_rows = flight_df[
    flight_df.duplicated()
]

In [36]:
print("Duplicate rows:", len(duplicate_rows))

Duplicate rows: 1922


In [37]:
duplicate_rows.nunique()

Loyalty Number                 162
Year                             2
Month                           12
Total Flights                    1
Distance                         1
Points Accumulated               1
Points Redeemed                  1
Dollar Cost Points Redeemed      1
dtype: int64

In [38]:
flight_df = flight_df.drop_duplicates()

In [39]:
merged_df = pd.merge(
    flight_df,
    loyalty_df,
    on="Loyalty Number",
    how="left"
)

In [40]:
merged_df["Date"] = pd.to_datetime(
    merged_df["Year"].astype(str) + "-" +
    merged_df["Month"].astype(str) + "-01"
)

In [41]:
merged_df = merged_df.sort_values([
    "Loyalty Number",
    "Date"
])

In [42]:
print(
    "Min Date:",
    merged_df["Date"].min()
)

print(
    "Max Date:",
    merged_df["Date"].max()
)

Min Date: 2017-01-01 00:00:00
Max Date: 2018-12-01 00:00:00


**Monthly Activity Distribution**

In [43]:
merged_df["Total Flights"].describe()

count    391014.000000
mean          1.301253
std           1.965387
min           0.000000
25%           0.000000
50%           0.000000
75%           2.000000
max          28.000000
Name: Total Flights, dtype: float64

**How many Zero-Activity Months?**

In [44]:
zero_activity_pct = (
    (merged_df["Total Flights"] == 0).mean()
) * 100

print(f"Zero Flight Months: {zero_activity_pct:.2f}%")

Zero Flight Months: 54.27%


**Active vs Inactive Customers**

In [45]:
customer_activity = merged_df.groupby(
    "Loyalty Number"
)["Total Flights"].sum()

print(customer_activity.describe())

count    16737.000000
mean        30.400191
std         16.865558
min          0.000000
25%         20.000000
50%         34.000000
75%         42.000000
max        106.000000
Name: Total Flights, dtype: float64


**Customers with Zero Flights Entirely**

In [46]:
zero_flight_customers = customer_summary[
    customer_summary["Total Flights"] == 0
]

zero_flight_customers[[
    "Loyalty Number",
    "CLV",
    "Loyalty Card",
    "Enrollment Type"
]].head(10)

,Loyalty Number,CLV,Loyalty Card,Enrollment Type
25,101356,7482.64,Nova,Standard
47,102308,12866.89,Star,Standard
64,103352,5555.49,Aurora,Standard
67,103383,13466.92,Aurora,Standard
91,104739,5498.94,Nova,Standard
97,105145,4839.11,Star,Standard
102,105352,3923.27,Star,Standard
105,105700,7944.97,Nova,Standard
123,106589,11638.90,Star,Standard
166,108481,11633.58,Aurora,Standard


In [47]:
zero_flight_customers["CLV"].describe()

count     1570.000000
mean      8360.820287
std       7537.885678
min       1994.770000
25%       4141.142500
50%       5907.950000
75%       9075.770000
max      83325.380000
Name: CLV, dtype: float64

In [48]:
true_inactive = zero_flight_customers[
    zero_flight_customers["CLV"] == 0
]

print("Customers with zero flights AND zero CLV:",
    len(true_inactive))

Customers with zero flights AND zero CLV: 0


In [49]:
historical_customers = zero_flight_customers[
    zero_flight_customers["CLV"] > 0
]

print("Customers with zero recent flights but positive CLV:",
    len(historical_customers))

Customers with zero recent flights but positive CLV: 1570


A substantial group of loyalty members retained positive historical CLV despite showing zero flight activity during the full 2017–2018 observation window, highlighting the disconnect between historical value and current engagement.

**Tier-wise Cancellation Rate**

In [50]:
loyalty_df["Cancelled"] = np.where(
    loyalty_df["Cancellation Year"].notnull(),
    1,
    0
)

loyalty_df.groupby("Loyalty Card")[
    "Cancelled"
].mean() * 100

Loyalty Card
Aurora    13.094197
Nova      12.643273
Star      11.797826
Name: Cancelled, dtype: float64

Cancellation rates were broadly similar across loyalty tiers, although Aurora members showed a slightly higher observed cancellation rate.

**Enrollment Type vs Cancellation**

In [51]:
loyalty_df.groupby("Enrollment Type")[
    "Cancelled"
].mean() * 100

Enrollment Type
2018 Promotion    11.843460
Standard          12.381073
Name: Cancelled, dtype: float64

No major difference in formal cancellation rates was observed between promotional and standard enrollment groups

### **Observations**
- Salary Missingness is NOT Random
  - Salary missingness depends on education category
- 54.49% months have zero flights
  - Most customers are inactive in most months
  - Monthly flight behaviour is sparse
  - Most people dont fly monthly
- Strong Evidence for Seasonal/Irregular Behaviour
  - Median 34 and max 106 flights
  - Many users do fly overall, just not every month
  - Airline customers are intermittent users
- 1570 Customers didnt Flew at all during 2017-18 (silent churn)
  - Dormant customers who were once active but completely disengaged during observation period
  - Reacquiring old valuable customers is cheaper than acquiring new ones.
  - Even premium users can disappear silently.
- Elite Users Churn Slightly MORE
  - High-value users are harder to retain
  - Maybe due to more competition, higher expectations or more airline alternatives.
- Promotion Users DONT Churn More
  - Promotional acquisiion may not be low quality

# **What does Disengagement actually look like**

**Monthly Activity per Customer**

In [112]:
customer_monthly_activity = (
    merged_df[merged_df["Total Flights"] > 0]
    .drop_duplicates(
        subset=["Loyalty Number", "Year", "Month"]
    )
    .groupby("Loyalty Number")
    .size()
    .reset_index(name="Active Months")
)

**Distribution of Active Months**

In [67]:
customer_monthly_activity.describe()

,Loyalty Number,Active Months
count,15167.000000,15167.000000
mean,549462.620294,11.734226
std,258599.674352,4.771680
min,100018.000000,1.000000
25%,327130.500000,9.000000
50%,549200.000000,13.000000
75%,772299.500000,15.000000
max,999986.000000,24.000000


**Customers with very Low Engagement**

In [78]:
low_activity = customer_monthly_activity[
    customer_monthly_activity["Active Months"] <= 2
]

print("Customers active <= 2 months:",
    len(low_activity))

Customers active <= 2 months: 832


**Redemption Behaviour Analysis**

In [69]:
redeemers = customer_summary[
    customer_summary["Points Redeemed"] > 0
]

non_redeemers = customer_summary[
    customer_summary["Points Redeemed"] == 0
]

print("Redeemers:", len(redeemers))
print("Non Redeemers:", len(non_redeemers))

Redeemers: 11487
Non Redeemers: 5250


In [70]:
print("Redeemer Avg CLV:",
    redeemers["CLV"].mean())

print("Non Redeemer Avg CLV:",
    non_redeemers["CLV"].mean())

Redeemer Avg CLV: 7954.285755201532
Non Redeemer Avg CLV: 8064.624923809524


**Loyalty Tier Distribution**

In [71]:
loyalty_df["Loyalty Card"].value_counts(normalize=True) * 100

Loyalty Card
Star      45.629444
Nova      33.883014
Aurora    20.487543
Name: proportion, dtype: float64

**Province-Level Customer Distribution**

In [72]:
loyalty_df["Province"].value_counts().head(10)

Province
Ontario             5404
British Columbia    4409
Quebec              3300
Alberta              969
Manitoba             658
New Brunswick        636
Nova Scotia          518
Saskatchewan         409
Newfoundland         258
Yukon                110
Name: count, dtype: int64

**Create Monthly Activity Flag**

In [73]:
merged_df["Active Month"] = np.where(
    merged_df["Total Flights"] > 0,
    1,
    0
)

**Create Rolling Inactivity**

In [74]:
merged_df["Inactive"] = np.where(
    merged_df["Total Flights"] == 0,
    1,
    0
)

**Consecutive Inactive Months**

In [75]:
merged_df["Inactive Streak"] = merged_df.groupby(
    "Loyalty Number"
)["Inactive"].cumsum()

**Potential Behavioral Churn**

In [76]:
merged_df["Behavioral Churn"] = np.where(
    merged_df["Inactive Streak"] >= 6,
    1,
    0
)

**Behavioral Churn Rate**

In [77]:
merged_df["Behavioral Churn"].mean() * 100

np.float64(53.21164971075204)

### **Observations**
- 15167 out of 16737 customers have at least one active month during 2017-2018
- Customers were active for an average of 11.73 months out of 24 indicating that airline engagement is intermittent rather than continuous.
- More than 54% of customer-month records contained zero flights indicating that most airline customers do not fly every month - inactivity for one month alone should not be treated as churn.
- Around 832 customers were active in less than 2 months. These represent weakly engaged members, promotion-driven signups, low retention users.
- Premium-tier customers generate value through higher-value trips rather than significantly higher trip frequency.
- Promotional acquisition did not necessarily produce lower-quality customers.
- Simple redemption participation alone may not strongly explain customer value.

In [79]:
merged_df = merged_df.sort_values([
    "Loyalty Number",
    "Date"
])

merged_df["Inactive"] = np.where(
    merged_df["Total Flights"] == 0,
    1,
    0
)

merged_df["Inactive Streak"] = (
    merged_df.groupby("Loyalty Number")["Inactive"]
    .transform(
        lambda x: x.groupby(
            (x != x.shift()).cumsum()
        ).cumsum()
    )
)

In [80]:
merged_df["Behavioral Churn"] = np.where(
    merged_df["Inactive Streak"] >= 6,
    1,
    0
)

In [81]:
merged_df["Behavioral Churn"].mean() * 100

np.float64(16.844921153718282)

Formal cancellation rate was around 12% and behavioral churn rate is 16.84%. (difference is 4.8%). This suggests that some customers disengage behaviorally without formally cancelling.

**Behavioral analytics detects at-risk customers earlier than cancellation records alone**

Behavioral churn was defined as six consecutive months without flight activity. This definition was chosen because airline engagement is naturally intermittent, making short inactivity periods insufficient indicators of disengagement.

## **Identifying WHICH customers behaviorally churn the most**

In [82]:
customer_churn = merged_df.groupby(
    "Loyalty Number"
)["Behavioral Churn"].max().reset_index()

In [83]:
customer_summary = customer_summary.merge(
    customer_churn,
    on="Loyalty Number",
    how="left"
)

**Overall Customer-Level Churn Rate**

In [84]:
customer_summary["Behavioral Churn"].mean() * 100

np.float64(34.31319830316066)

**Churn by Loyalty Tier**

In [85]:
customer_summary.groupby("Loyalty Card")[
    "Behavioral Churn"
].mean() * 100

Loyalty Card
Aurora    34.966463
Nova      34.755775
Star      33.691240
Name: Behavioral Churn, dtype: float64

**Churn by Enrollment Type**

In [86]:
customer_summary.groupby("Enrollment Type")[
    "Behavioral Churn"
].mean() * 100

Enrollment Type
2018 Promotion     7.209063
Standard          35.982494
Name: Behavioral Churn, dtype: float64

2018 Promotion users showed substantially lower behavioral churn; however, this may partially reflect shorter observation duration due to later enrollment timing.

**Churn by Redemption Behavior**

In [87]:
customer_summary["Redeemer"] = np.where(
    customer_summary["Points Redeemed"] > 0,
    1,
    0
)

In [88]:
customer_summary.groupby("Redeemer")[
    "Behavioral Churn"
].mean() * 100

Redeemer
0    66.609524
1    19.552538
Name: Behavioral Churn, dtype: float64

**Churn by Active Months**

In [89]:
customer_summary = customer_summary.merge(
    customer_monthly_activity,
    on="Loyalty Number",
    how="left"
)

customer_summary["Active Months"] = customer_summary[
    "Active Months"
].fillna(0)

In [90]:
customer_summary.groupby("Behavioral Churn")[
    "Active Months"
].mean()

Behavioral Churn
0    13.767964
1     4.633119
Name: Active Months, dtype: float64

**CLV Bands**

In [91]:
customer_summary["CLV Segment"] = pd.qcut(
    customer_summary["CLV"],
    q=4,
    labels=[
        "Low Value",
        "Medium Value",
        "High Value",
        "Premium Value"
    ]
)

**Churn by CLV Segment**

In [92]:
customer_summary.groupby("CLV Segment")[
    "Behavioral Churn"
].mean() * 100

C:\Users\abc\AppData\Local\Temp\ipykernel_11584\3051973351.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  customer_summary.groupby("CLV Segment")[


CLV Segment
Low Value        33.850932
Medium Value     34.661260
High Value       34.554598
Premium Value    34.185991
Name: Behavioral Churn, dtype: float64

**Most Dangerous Customers - high CLV + churned customers**

In [95]:
at_risk_premium = customer_summary[
    (customer_summary["Behavioral Churn"] == 1) &
    (customer_summary["CLV Segment"] == "Premium Value")
]

at_risk_premium.sort_values(
    by="CLV",
    ascending=False
).head(10)

,Loyalty Number,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed,CLV,Loyalty Card,Enrollment Type,Redemption Ratio,Behavioral Churn,Redeemer,Active Months,CLV Segment
10297,652627,0,0,0.0,0,0,83325.38,Star,Standard,0.000000,1,0,0.0,Premium Value
13888,844145,0,0,0.0,0,0,74228.52,Nova,Standard,0.000000,1,0,0.0,Premium Value
12468,767366,8,15661,15661.0,504,91,73225.96,Aurora,Standard,0.032182,1,1,4.0,Premium Value
13776,838263,0,0,0.0,0,0,67907.27,Aurora,Standard,0.000000,1,0,0.0,Premium Value
1465,179870,33,48480,48480.0,0,0,66025.75,Aurora,Standard,0.000000,1,0,11.0,Premium Value
5998,424337,37,44466,44466.0,0,0,64618.76,Nova,Standard,0.000000,1,0,14.0,Premium Value
9171,592992,21,31945,31945.0,0,0,61850.19,Aurora,Standard,0.000000,1,0,7.0,Premium Value
1155,163405,4,4448,4448.0,0,0,60556.19,Nova,Standard,0.000000,1,0,2.0,Premium Value
2370,228030,12,14248,14248.0,0,0,58753.88,Aurora,Standard,0.000000,1,0,4.0,Premium Value
1364,173446,31,47727,47727.0,398,72,58207.13,Nova,Standard,0.008339,1,1,10.0,Premium Value


**Create Engagement Score**

In [96]:
customer_summary["Engagement Score"] = (
    customer_summary["Active Months"] * 0.4 +
    customer_summary["Total Flights"] * 0.3 +
    customer_summary["Points Accumulated"] / 10000 * 0.3
)

**Compare Engagement by Churn**

In [97]:
customer_summary.groupby("Behavioral Churn")[
    "Engagement Score"
].mean()

Behavioral Churn
0    19.320132
1     6.150538
Name: Engagement Score, dtype: float64

**Find Silent High-Value Customers**

In [100]:
silent_high_value = customer_summary[
    (customer_summary["CLV Segment"] == "Premium Value") &
    (customer_summary["Behavioral Churn"] == 1)
]

silent_high_value.head()

,Loyalty Number,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed,CLV,Loyalty Card,Enrollment Type,Redemption Ratio,Behavioral Churn,Redeemer,Active Months,CLV Segment,Engagement Score
38,101768,1,1733,1733.0,0,0,13916.93,Nova,Standard,0.0,1,0,1.0,Premium Value,0.75199
40,101895,27,47305,47305.0,0,0,24543.20,Nova,Standard,0.0,1,0,13.0,Premium Value,14.71915
47,102308,0,0,0.0,0,0,12866.89,Star,Standard,0.0,1,0,0.0,Premium Value,0.00000
54,102740,5,8927,8927.0,0,0,16922.13,Star,Standard,0.0,1,0,2.0,Premium Value,2.56781
67,103383,0,0,0.0,0,0,13466.92,Aurora,Standard,0.0,1,0,0.0,Premium Value,0.00000


### **Observations**
- Customer-Level Behavioral Churn = 34.3%
  - 34.3% of the customers experienced at least one 6-month inactivity streak.
  - Disengagement happens long before formal cancellation.
- Behavioral churn rates were broadly similar accross loyalty tiers.
- Promotion users are much more behaviorally active - atleast during observation window. But may be misleading as these users joined in 2018.
- Redemption is STRONGLY associated with Retention.
  - Customers who redeem points are dramatically more engaged.
- Sustained engagement consistency is highly predictive of retention.
- CLV does not explain churn - challenges Traditional CLV Thinking
- Silent high value customers exist.
  - Historically generated high value but are now disengaging silently.

An initial engagement score was constructed using weighted behavioral indicators to approximate customer engagement intensity.

**Create Customer Tenure**

In [101]:
customer_summary["Tenure"] = (
    2018 - merged_df.groupby("Loyalty Number")[
        "Enrollment Year"
    ].first().values
)

In [102]:
enrollment_info = loyalty_df[[
    "Loyalty Number",
    "Enrollment Year"
]]

customer_summary = customer_summary.merge(
    enrollment_info,
    on="Loyalty Number",
    how="left"
)

customer_summary["Tenure"] = (
    2018 - customer_summary["Enrollment Year"]
)

In [103]:
customer_summary.groupby("Enrollment Year")[
    "Behavioral Churn"
].mean() * 100

Enrollment Year
2012    18.920522
2013    18.439716
2014    18.902954
2015    20.806521
2016    18.159609
2017    69.159630
2018    62.558140
Name: Behavioral Churn, dtype: float64

In [104]:
customer_summary.groupby("Tenure")[
    "Behavioral Churn"
].mean() * 100

Tenure
0    62.558140
1    69.159630
2    18.159609
3    20.806521
4    18.902954
5    18.439716
6    18.920522
Name: Behavioral Churn, dtype: float64

In [105]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()

engagement_features = customer_summary[[
    "Active Months",
    "Total Flights",
    "Points Accumulated"
]]

scaled_features = scaler.fit_transform(
    engagement_features
)

In [106]:
customer_summary["Engagement Score"] = (
    scaled_features[:,0] +
    scaled_features[:,1] +
    scaled_features[:,2]
) / 3

## **PROBLEM!**
Using churn definition as 6 inactive months and not considering enrollment year.

## **REBUILD CHURN ANALYSIS**

In [146]:
loyalty_df["Enrollment Date"] = pd.to_datetime(
    loyalty_df["Enrollment Year"].astype(str) + "-" +
    loyalty_df["Enrollment Month"].astype(str) + "-01"
)

In [147]:
flight_df["Date"] = pd.to_datetime(
    flight_df["Year"].astype(str) + "-" +
    flight_df["Month"].astype(str) + "-01"
)

In [148]:
merged_df = pd.merge(
    flight_df,
    loyalty_df,
    on="Loyalty Number",
    how="left"
)

In [149]:
merged_df = merged_df[
    merged_df["Date"] >= merged_df["Enrollment Date"]
]

In [150]:
merged_df = merged_df.sort_values([
    "Loyalty Number",
    "Date"
])

In [151]:
merged_df["Inactive"] = np.where(
    merged_df["Total Flights"] == 0,
    1,
    0
)

In [152]:
merged_df["Inactive Streak"] = (
    merged_df.groupby("Loyalty Number")["Inactive"]
    .transform(
        lambda x: x.groupby(
            (x != x.shift()).cumsum()
        ).cumsum()
    )
)

In [153]:
merged_df["Behavioral Churn"] = np.where(
    merged_df["Inactive Streak"] >= 6,
    1,
    0
)

In [154]:
customer_churn = merged_df.groupby(
    "Loyalty Number"
)["Behavioral Churn"].max().reset_index()

In [155]:
customer_summary = merged_df.groupby(
    "Loyalty Number"
).agg({
    "Total Flights": "sum",
    "Distance": "sum",
    "Points Accumulated": "sum",
    "Points Redeemed": "sum",
    "Dollar Cost Points Redeemed": "sum",
    "CLV": "first",
    "Loyalty Card": "first",
    "Enrollment Type": "first",
    "Enrollment Year": "first"
}).reset_index()

In [156]:
customer_summary = customer_summary.merge(
    customer_churn,
    on="Loyalty Number",
    how="left"
)

In [157]:
customer_monthly_activity = (
    merged_df[merged_df["Total Flights"] > 0]
    .drop_duplicates(
        subset=["Loyalty Number", "Year", "Month"]
    )
    .groupby("Loyalty Number")
    .size()
    .reset_index(name="Active Months")
)

In [158]:
customer_summary = customer_summary.merge(
    customer_monthly_activity,
    on="Loyalty Number",
    how="left"
)

In [159]:
customer_summary["Active Months"] = customer_summary[
    "Active Months"
].fillna(0)

In [160]:
customer_summary["Redemption Ratio"] = np.where(
    customer_summary["Points Accumulated"] > 0,
    customer_summary["Points Redeemed"] /
    customer_summary["Points Accumulated"],
    0
)

In [161]:
customer_summary["Redeemer"] = np.where(
    customer_summary["Points Redeemed"] > 0,
    1,
    0
)

In [162]:
customer_summary["CLV Segment"] = pd.qcut(
    customer_summary["CLV"],
    q=4,
    labels=[
        "Low Value",
        "Medium Value",
        "High Value",
        "Premium Value"
    ]
)

In [163]:
customer_summary["Tenure"] = (
    2018 - customer_summary["Enrollment Year"]
)

In [164]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()

scaled_features = scaler.fit_transform(
    customer_summary[[
        "Active Months",
        "Total Flights",
        "Points Accumulated"
    ]]
)

customer_summary["Engagement Score"] = (
    scaled_features[:,0] +
    scaled_features[:,1] +
    scaled_features[:,2]
) / 3

In [165]:
customer_summary.head()

,Loyalty Number,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed,CLV,Loyalty Card,Enrollment Type,Enrollment Year,Behavioral Churn,Active Months,Redemption Ratio,Redeemer,CLV Segment,Tenure,Engagement Score
0,100018,46,81190,81190.0,1513,272,7919.20,Aurora,Standard,2016,0,18.0,0.018635,1,High Value,2,0.501201
1,100102,51,68918,68918.0,1195,215,2887.74,Nova,Standard,2013,0,17.0,0.017339,1,Low Value,5,0.488405
2,100140,47,72856,72856.0,593,107,2838.07,Nova,Standard,2016,0,17.0,0.008139,1,Low Value,2,0.480226
3,100214,22,38236,38236.0,861,155,4170.57,Star,Standard,2015,1,9.0,0.022518,1,Medium Value,3,0.244402
4,100272,37,54997,54997.0,1007,182,6622.05,Star,Standard,2014,0,13.0,0.018310,1,High Value,4,0.369802


In [166]:
customer_summary.columns

Index(['Loyalty Number', 'Total Flights', 'Distance', 'Points Accumulated',
       'Points Redeemed', 'Dollar Cost Points Redeemed', 'CLV', 'Loyalty Card',
       'Enrollment Type', 'Enrollment Year', 'Behavioral Churn',
       'Active Months', 'Redemption Ratio', 'Redeemer', 'CLV Segment',
       'Tenure', 'Engagement Score'],
      dtype='object')

In [167]:
customer_summary.describe()

,Loyalty Number,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed,CLV,Enrollment Year,Behavioral Churn,Active Months,Redemption Ratio,Redeemer,Tenure,Engagement Score
count,16737.000000,16737.000000,16737.000000,16737.000000,16737.000000,16737.000000,16737.000000,16737.000000,16737.000000,16737.000000,16737.000000,16737.000000,16737.000000,16737.000000
mean,549735.880445,29.907212,44837.507439,46828.866912,720.934098,129.788971,7988.896536,2015.253211,0.157316,10.442373,0.016138,0.676406,2.746789,0.300951
std,258912.132453,17.005076,26288.413210,30510.107462,714.296028,128.589565,6860.982280,1.979111,0.364109,5.754709,0.026520,0.467861,1.979111,0.161674
min,100018.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1898.010000,2012.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,326603.000000,18.000000,26520.000000,26771.000000,0.000000,0.000000,3980.840000,2014.000000,0.000000,6.000000,0.000000,0.000000,1.000000,0.190429
50%,550434.000000,33.000000,48806.000000,49306.000000,552.000000,99.000000,5780.180000,2015.000000,0.000000,12.000000,0.011588,1.000000,3.000000,0.345216
75%,772019.000000,42.000000,62563.000000,63457.000000,1152.000000,208.000000,8940.580000,2017.000000,0.000000,15.000000,0.023303,1.000000,4.000000,0.419110
max,999986.000000,102.000000,178858.000000,268287.000000,4479.000000,806.000000,83325.380000,2018.000000,1.000000,24.000000,1.194352,1.000000,6.000000,0.808097


In [168]:
customer_summary["Behavioral Churn"].mean() * 100

np.float64(15.731612594849734)

In [169]:
customer_summary.groupby("Loyalty Card")[
    "Behavioral Churn"
].mean() * 100

Loyalty Card
Aurora    17.002041
Nova      16.028919
Star      14.940422
Name: Behavioral Churn, dtype: float64

In [170]:
customer_summary.groupby("Enrollment Year")[
    "Behavioral Churn"
].mean() * 100

Enrollment Year
2012    18.920522
2013    18.439716
2014    18.902954
2015    20.806521
2016    18.159609
2017    15.440290
2018     3.621262
Name: Behavioral Churn, dtype: float64

In [171]:
customer_summary.groupby("Redeemer")[
    "Behavioral Churn"
].mean() * 100

Redeemer
0    34.028804
1     6.978182
Name: Behavioral Churn, dtype: float64

In [172]:
customer_summary.groupby("Behavioral Churn")[
    "Active Months"
].mean()

Behavioral Churn
0    11.711429
1     3.644512
Name: Active Months, dtype: float64

In [173]:
customer_summary.groupby("Behavioral Churn")[
    "Engagement Score"
].mean()

Behavioral Churn
0    0.338243
1    0.101194
Name: Engagement Score, dtype: float64

In [174]:
customer_summary.groupby("CLV Segment")[
    "Behavioral Churn"
].mean() * 100

C:\Users\abc\AppData\Local\Temp\ipykernel_11584\3051973351.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  customer_summary.groupby("CLV Segment")[


CLV Segment
Low Value        14.906832
Medium Value     16.555344
High Value       15.684866
Premium Value    15.778150
Name: Behavioral Churn, dtype: float64

In [175]:
silent_high_value = customer_summary[
    (customer_summary["Behavioral Churn"] == 1) &
    (customer_summary["CLV Segment"] == "Premium Value")
]

silent_high_value.sort_values(
    by="CLV",
    ascending=False
).head(10)

,Loyalty Number,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed,CLV,Loyalty Card,Enrollment Type,Enrollment Year,Behavioral Churn,Active Months,Redemption Ratio,Redeemer,CLV Segment,Tenure,Engagement Score
10297,652627,0,0,0.0,0,0,83325.38,Star,Standard,2015,1,0.0,0.000000,0,Premium Value,3,0.000000
13888,844145,0,0,0.0,0,0,74228.52,Nova,Standard,2015,1,0.0,0.000000,0,Premium Value,3,0.000000
12468,767366,8,15661,15661.0,504,91,73225.96,Aurora,Standard,2017,1,4.0,0.032182,1,Premium Value,1,0.101157
13776,838263,0,0,0.0,0,0,67907.27,Aurora,Standard,2012,1,0.0,0.000000,0,Premium Value,6,0.000000
9171,592992,21,31945,31945.0,0,0,61850.19,Aurora,Standard,2013,1,7.0,0.000000,0,Premium Value,5,0.205540
1364,173446,31,47727,47727.0,398,72,58207.13,Nova,Standard,2017,1,10.0,0.008339,1,Premium Value,1,0.299495
10672,672917,0,0,0.0,0,0,58166.55,Star,Standard,2015,1,0.0,0.000000,0,Premium Value,3,0.000000
4513,344232,34,46704,46704.0,1094,197,51426.25,Star,Standard,2012,1,13.0,0.023424,1,Premium Value,6,0.349694
7976,529331,0,0,0.0,0,0,49221.43,Aurora,Standard,2013,1,0.0,0.000000,0,Premium Value,5,0.000000
11873,737027,0,0,0.0,0,0,49221.43,Nova,Standard,2012,1,0.0,0.000000,0,Premium Value,6,0.000000


### **Observations**

* Final Customer-Level Behavioral Churn Rate:

  * After correcting for enrollment timing and removing pre-enrollment inactivity bias, the final behavioral churn rate was: 15.7%
  * This is higher than formal cancellation rates, indicating that many customers disengage behaviorally before formally leaving the loyalty program.

---

* Behavioral churn rates were broadly similar across loyalty tiers.

  * Aurora, Nova, and Star members showed only minor differences in churn behavior.
  * Loyalty tier alone does not appear to strongly predict disengagement.

---

* Enrollment timing significantly affects churn interpretation.

  * Customers enrolled in 2018 showed very low churn rates.
  * However, these users had a shorter observation window and less opportunity to accumulate long inactivity streaks.
  * This highlights the importance of considering enrollment date while defining churn and avoiding temporal bias.

---

* Redemption behavior is strongly associated with retention.

  * Customers who redeemed loyalty points had substantially lower behavioral churn rates compared to non-redeemers.
  * Non-redeemers showed nearly 5 times higher churn risk.
  * Redemption participation appears to strengthen engagement with the airline ecosystem.

---

* Sustained engagement consistency is highly predictive of retention.

  * Customers with higher active-month counts were significantly less likely to churn.
  * Churned users had far fewer active months on average, indicating that consistent interaction matters more than occasional high activity.

---

* The normalized Engagement Score strongly separated active and churned customers.

  * Customers with low engagement scores were far more likely to behaviorally churn.
  * This validates the usefulness of combining behavioral signals such as:

    * active months
    * flight frequency
    * points accumulation

---

* Historical CLV alone does not explain future disengagement.

  * Behavioral churn rates remained relatively similar across CLV segments.
  * This challenges the traditional assumption that historically valuable customers are automatically loyal going forward.

---

* Silent high-value customers exist within the loyalty base.

  * Several customers with very high historical CLV showed:

    * zero recent flights
    * zero active months
    * behavioral churn signals
  * These customers represent hidden revenue-risk segments that traditional loyalty dashboards may fail to detect.

---

* A portion of customers showed positive CLV despite zero activity during 2017–2018.

  * This suggests that CLV reflects historical lifetime contribution rather than recent engagement.
  * Therefore, relying only on CLV may hide recently disengaged customers.

---

* More than half of monthly customer records contained zero flights.

  * Airline engagement is naturally intermittent.
  * Therefore, short inactivity periods should not automatically be interpreted as churn.
  * This justified defining behavioral churn using: 6 consecutive inactive months instead of single-month inactivity.


## **Define Business-Driven Behavioral Segments**

**Create Initial Rule-Based Segments**

In [176]:
conditions = [

    # Silent High Value
    (
        (customer_summary["CLV Segment"] == "Premium Value") &
        (customer_summary["Behavioral Churn"] == 1)
    ),

    # Engaged Loyalists
    (
        (customer_summary["Behavioral Churn"] == 0) &
        (customer_summary["Engagement Score"] >= 0.5)
    ),

    # Reward-Oriented Active Users
    (
        (customer_summary["Redeemer"] == 1) &
        (customer_summary["Behavioral Churn"] == 0)
    ),

    # Dormant Low Engagement
    (
        (customer_summary["Behavioral Churn"] == 1) &
        (customer_summary["Engagement Score"] < 0.15)
    ),

    # New Promotion Users
    (
        (customer_summary["Enrollment Type"] == "2018 Promotion")
    )
]

segment_names = [
    "Silent High Value",
    "Engaged Loyalists",
    "Reward-Oriented Active",
    "Dormant Low Engagement",
    "New Promotion Users"
]

customer_summary["Customer Segment"] = np.select(
    conditions,
    segment_names,
    default="Regular Customers"
)

**Segment Distribution**

In [177]:
customer_summary["Customer Segment"].value_counts()

Customer Segment
Reward-Oriented Active    9700
Regular Customers         3708
Dormant Low Engagement    1350
Engaged Loyalists          948
Silent High Value          660
New Promotion Users        371
Name: count, dtype: int64

**Segment Percentages**

In [178]:
customer_summary["Customer Segment"].value_counts(
    normalize=True
) * 100

Customer Segment
Reward-Oriented Active    57.955428
Regular Customers         22.154508
Dormant Low Engagement     8.065962
Engaged Loyalists          5.664098
Silent High Value          3.943359
New Promotion Users        2.216646
Name: proportion, dtype: float64

**Segment-Level Metrics**

In [179]:
customer_summary.groupby("Customer Segment")[[
    "CLV",
    "Total Flights",
    "Active Months",
    "Engagement Score",
    "Behavioral Churn"
]].mean().sort_values(
    by="CLV",
    ascending=False
)

,CLV,Total Flights,Active Months,Engagement Score,Behavioral Churn
Customer Segment,,,,,
Silent High Value,17239.017530,9.440909,3.454545,0.096386,1.000000
New Promotion Users,8115.503558,42.700809,5.188679,0.327574,0.002695
Reward-Oriented Active,7970.989566,35.076082,12.766082,0.358656,0.000000
Engaged Loyalists,7967.569947,58.484177,16.842827,0.549066,0.000000
Regular Customers,7419.957737,21.411003,7.935545,0.220116,0.167745
Dormant Low Engagement,5138.147807,2.526667,0.997037,0.026826,1.000000


# **Customer Segmentation Analysis**

## 1. Reward-Oriented Active Users
### Segment Size
~58% of customers

### Characteristics
- High redemption activity
- Consistent engagement
- Moderate-to-high flight frequency
- Strong retention behavior
- Behavioral churn almost negligible

### Key Metrics
- Avg Flights: 35
- Avg Active Months: 12.8
- Avg Engagement Score: 0.36
- Avg CLV: 7971

### Interpretation
These customers actively participate in the loyalty ecosystem and appear highly engaged with the rewards structure. They represent the healthiest and most stable customer base.

### Business Strategy
- Introduce milestone-based rewards
- Offer bonus redemption campaigns
- Expand partner reward ecosystem
- Encourage referrals and upselling

---

## 2. Engaged Loyalists
### Segment Size
~5.7% of customers

### Characteristics
- Highest flight frequency
- Highest engagement score
- Consistently active across months
- Extremely low churn

### Key Metrics
- Avg Flights: 58
- Avg Active Months: 16.8
- Avg Engagement Score: 0.55
- Avg CLV: 7968

### Interpretation
These are highly committed and behaviorally loyal customers, likely representing frequent travelers or business travelers.

### Business Strategy
- Provide VIP treatment and premium support
- Offer lounge access and priority services
- Introduce exclusive elite-tier benefits
- Protect retention aggressively

---

## 3. Silent High Value Customers
### Segment Size
~3.9% of customers

### Characteristics
- Very high historical CLV
- Extremely low recent engagement
- High behavioral churn
- Low recent activity

### Key Metrics
- Avg CLV: 17239
- Avg Flights: 9.4
- Avg Active Months: 3.5
- Avg Engagement Score: 0.096
- Behavioral Churn: 100%

### Interpretation
These customers historically generated high value but are now silently disengaging. Traditional CLV-focused systems may still classify them as valuable despite clear behavioral decline.

### Business Strategy
- Launch proactive win-back campaigns
- Offer personalized route discounts
- Use premium retention outreach
- Trigger early recovery interventions before formal churn

---

## 4. Dormant Low Engagement Users
### Segment Size
~8.1% of customers

### Characteristics
- Very low activity
- Minimal engagement
- Extremely high churn
- Weak loyalty behavior

### Key Metrics
- Avg Flights: 2.5
- Avg Active Months: 1
- Avg Engagement Score: 0.027
- Avg CLV: 5138
- Behavioral Churn: 100%

### Interpretation
These customers show weak attachment to the airline and may represent low-retention or low-frequency users.

### Business Strategy
- Use low-cost automated reactivation campaigns
- Send occasional discount reminders
- Avoid expensive retention spending
- Focus on scalable digital outreach

---

## 5. New Promotion Users
### Segment Size
~2.2% of customers

### Characteristics
- Recently acquired customers
- High initial activity
- Very low churn during observation period

### Key Metrics
- Avg Flights: 42.7
- Avg Active Months: 5.2
- Avg Engagement Score: 0.33
- Avg CLV: 8116

### Interpretation
These users appear highly active initially, although their observation window is shorter due to recent enrollment. Their long-term retention behavior remains uncertain.

### Business Strategy
- Encourage first reward redemption quickly
- Build travel habits early
- Introduce onboarding journeys
- Use personalized engagement nudges

---

## 6. Regular Customers
### Segment Size
~22.2% of customers

### Characteristics
- Moderate activity levels
- Average engagement
- Moderate churn probability

### Key Metrics
- Avg Flights: 21
- Avg Active Months: 7.9
- Avg Engagement Score: 0.22
- Avg CLV: 7420
- Behavioral Churn: 16.8%

### Interpretation
These customers form the middle segment of the loyalty base and are neither strongly loyal nor fully disengaged.

### Business Strategy
- Apply predictive churn monitoring
- Use personalized offers
- Encourage increased engagement frequency
- Promote loyalty participation

---

# Overall Segmentation Insights

- Behavioral segmentation produced more actionable customer groups than loyalty tiers alone.
- Historical CLV alone was insufficient to identify disengagement risk.
- Redemption participation emerged as one of the strongest indicators of retention.
- Silent high-value customers represent hidden revenue risk that traditional loyalty systems may fail to detect.
- Engagement consistency was more predictive of retention than historical customer value alone.
- Different customer groups require different retention approaches instead of generic loyalty campaigns.

## **CHURN PREDICTION MODEL**

**Whether customer behaviorally churns during Jul–Dec 2018**

In [180]:
cutoff_date = "2018-07-01"

In [181]:
historical_df = merged_df[
    merged_df["Date"] < cutoff_date
]

future_df = merged_df[
    merged_df["Date"] >= cutoff_date
]

In [182]:
future_df = future_df.sort_values([
    "Loyalty Number",
    "Date"
])

In [183]:
future_df["Inactive"] = np.where(
    future_df["Total Flights"] == 0,
    1,
    0
)

In [184]:
future_df["Inactive Streak"] = (
    future_df.groupby("Loyalty Number")["Inactive"]
    .transform(
        lambda x: x.groupby(
            (x != x.shift()).cumsum()
        ).cumsum()
    )
)

In [185]:
future_df["Future Churn"] = np.where(
    future_df["Inactive Streak"] >= 6,
    1,
    0
)

In [186]:
future_churn = future_df.groupby(
    "Loyalty Number"
)["Future Churn"].max().reset_index()

In [187]:
model_features = historical_df.groupby(
    "Loyalty Number"
).agg({
    "Total Flights": "sum",
    "Distance": "sum",
    "Points Accumulated": "sum",
    "Points Redeemed": "sum",
    "Dollar Cost Points Redeemed": "sum",
    "CLV": "first",
    "Loyalty Card": "first",
    "Enrollment Type": "first",
    "Enrollment Year": "first"
}).reset_index()

In [188]:
historical_active_months = (
    historical_df[historical_df["Total Flights"] > 0]
    .drop_duplicates(
        subset=["Loyalty Number", "Year", "Month"]
    )
    .groupby("Loyalty Number")
    .size()
    .reset_index(name="Active Months")
)

In [189]:
model_df = model_features.merge(
    historical_active_months,
    on="Loyalty Number",
    how="left"
)

In [190]:
model_df = model_df.merge(
    future_churn,
    on="Loyalty Number",
    how="left"
)

In [191]:
model_df["Active Months"] = model_df[
    "Active Months"
].fillna(0)

model_df["Future Churn"] = model_df[
    "Future Churn"
].fillna(0)

In [192]:
model_df["Redeemer"] = np.where(
    model_df["Points Redeemed"] > 0,
    1,
    0
)

In [193]:
model_df["Redemption Ratio"] = np.where(
    model_df["Points Accumulated"] > 0,
    model_df["Points Redeemed"] /
    model_df["Points Accumulated"],
    0
)

In [194]:
model_df["Tenure"] = (
    2018 - model_df["Enrollment Year"]
)

In [196]:
model_df["Future Churn"].value_counts()

Future Churn
0    13259
1     2147
Name: count, dtype: int64

In [197]:
model_df["Future Churn"].mean() * 100

np.float64(13.936128780994416)

In [198]:
features = [
    "Total Flights",
    "Distance",
    "Points Accumulated",
    "Points Redeemed",
    "CLV",
    "Active Months",
    "Redeemer",
    "Redemption Ratio",
    "Tenure"
]

X = model_df[features]

y = model_df["Future Churn"]

In [199]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

**Train Baseline Model using Random Forest**

In [200]:
from sklearn.ensemble import RandomForestClassifier

rf_model = RandomForestClassifier(
    n_estimators=100,
    random_state=42
)

rf_model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [201]:
y_pred = rf_model.predict(X_test)

In [202]:
from sklearn.metrics import classification_report

print(classification_report(
    y_test,
    y_pred
))

              precision    recall  f1-score   support

           0       0.95      0.99      0.97      2652
           1       0.90      0.68      0.78       430

    accuracy                           0.94      3082
   macro avg       0.92      0.84      0.87      3082
weighted avg       0.94      0.94      0.94      3082



In [203]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred)

array([[2618,   34],
       [ 136,  294]])

In [204]:
importance_df = pd.DataFrame({
    "Feature": features,
    "Importance": rf_model.feature_importances_
})

importance_df.sort_values(
    by="Importance",
    ascending=False
)

,Feature,Importance
2,Points Accumulated,0.230576
1,Distance,0.198534
8,Tenure,0.184161
5,Active Months,0.112998
0,Total Flights,0.102077
4,CLV,0.099499
7,Redemption Ratio,0.035949
3,Points Redeemed,0.032536
6,Redeemer,0.003671


# Predictive Churn Modeling

## Objective
The goal of the predictive model was to identify customers likely to behaviorally churn in the future using only historical behavioral information available before the prediction period.

To avoid data leakage, future customer activity was strictly excluded from feature generation.

---

# Modeling Framework

## Temporal Split

| Period | Purpose |
|---|---|
| Jan 2017 – Jun 2018 | Historical feature generation |
| Jul 2018 – Dec 2018 | Future churn prediction window |

This structure ensured the model mimicked a realistic business scenario where predictions are made using only past information.

---

# Churn Definition

Behavioral churn was defined as:

\[
6
\]
consecutive months with zero flight activity.

:contentReference[oaicite:0]{index=0}

This definition was selected because airline travel is naturally intermittent, and shorter inactivity periods may not necessarily indicate disengagement.

---

# Target Variable

The model predicted whether a customer would experience behavioral churn during the future observation window (Jul–Dec 2018).

Final Future Churn Rate:

\[
13.9\%
\]

:contentReference[oaicite:1]{index=1}

---

# Features Used

The following behavioral and customer-level features were used:

- Total Flights
- Distance Traveled
- Points Accumulated
- Points Redeemed
- CLV
- Active Months
- Redeemer Flag
- Redemption Ratio
- Customer Tenure

---

# Model Selection

A Random Forest Classifier was used because it:
- handles nonlinear behavioral relationships well
- provides feature importance interpretation
- performs strongly on tabular business datasets
- balances predictive performance with explainability

---

# Model Performance

| Metric | Churn Class Performance |
|---|---|
| Precision | 0.90 |
| Recall | 0.68 |
| F1 Score | 0.78 |
| Overall Accuracy | 94% |

---

# Performance Interpretation

## High Precision
The model achieved:

\[
90\%
\]
precision for churn prediction.

:contentReference[oaicite:2]{index=2}

This means when the model predicts a customer is likely to churn, it is correct most of the time.

This is highly valuable for airline retention campaigns because:
- marketing interventions are expensive
- false targeting reduces campaign efficiency
- retention resources should focus on high-confidence customers

---

## Recall Performance

The model captured approximately:

\[
68\%
\]
of actual churners.

:contentReference[oaicite:3]{index=3}

While some churners remain undetected, the model successfully identifies the majority of disengaging customers.

---

# Confusion Matrix Interpretation

| | Predicted Stay | Predicted Churn |
|---|---|---|
| Actual Stay | 2618 | 34 |
| Actual Churn | 136 | 294 |

Key observation:
- False positives were very low.
- Only 34 loyal customers were incorrectly flagged as churn risks.

This makes the model practical for real-world marketing deployment.

---

# Feature Importance Analysis

| Feature | Importance |
|---|---|
| Points Accumulated | 0.23 |
| Distance | 0.20 |
| Tenure | 0.18 |
| Active Months | 0.11 |
| Total Flights | 0.10 |
| CLV | 0.10 |
| Redemption Ratio | 0.04 |
| Points Redeemed | 0.03 |
| Redeemer Flag | 0.004 |

---

# Key Predictive Insights

## Behavioral Signals Dominated Prediction
The strongest predictors of churn were:
- points accumulation
- travel distance
- active engagement duration
- customer tenure

This suggests that ongoing behavioral engagement is more informative than static demographic or loyalty-tier information.

---

## Historical CLV Was Not a Strong Predictor
CLV showed relatively low predictive importance compared to behavioral variables.

This challenges traditional loyalty assumptions that historically valuable customers are automatically loyal in the future.

Instead, the analysis showed that:
- high historical value does not guarantee future engagement
- active participation and consistent interaction matter more

---

## Engagement Consistency Was Critical
Customers with:
- more active months
- higher activity frequency
- stronger ongoing participation

were substantially less likely to churn.

This validates the importance of engagement-based loyalty intelligence.

---

# Strategic Business Implications

The predictive model demonstrates that airlines can improve retention intelligence by shifting from:
- static customer valuation
to:
- dynamic behavioral monitoring

Traditional loyalty systems often focus heavily on:
- loyalty tiers
- accumulated value
- cancellation records

However, this analysis showed that:
- silent disengagement occurs before formal cancellation
- behavioral inactivity provides earlier warning signals
- retention actions should be triggered proactively

---

# Final Strategic Conclusion

The predictive churn framework successfully identified at-risk customers using behavioral signals while avoiding future data leakage.

The strongest business insight from the model was:

> Customer engagement behavior predicts future loyalty more effectively than historical customer value alone.

This creates opportunities for airlines to:
- detect silent churn earlier
- personalize retention interventions
- prioritize high-risk high-value customers
- optimize loyalty program effectiveness

In [205]:
customer_summary["Recommended Action"] = np.select(
    [
        customer_summary["Customer Segment"] == "Silent High Value",
        customer_summary["Customer Segment"] == "Dormant Low Engagement",
        customer_summary["Customer Segment"] == "Reward-Oriented Active",
        customer_summary["Customer Segment"] == "Engaged Loyalists",
        customer_summary["Customer Segment"] == "New Promotion Users"
    ],
    [
        "Premium Win-Back Campaign",
        "Automated Reactivation Email",
        "Loyalty Expansion Campaign",
        "VIP Retention Program",
        "Onboarding & Habit Formation"
    ],
    default="Standard Retention Strategy"
)

In [206]:
customer_summary.to_csv(
    "final_airline_dashboard_data.csv",
    index=False
)

In [207]:
customer_summary.to_csv("customer_summary.csv", index=False)

In [208]:
segment_metrics = customer_summary.groupby("Customer Segment")[[
    "CLV", "Total Flights", "Active Months", "Engagement Score", "Behavioral Churn"
]].mean().reset_index()
segment_metrics["Count"] = customer_summary.groupby("Customer Segment").size().values
segment_metrics.to_csv("segment_metrics.csv", index=False)